### DISCUSSION TITLE
Can MOOSE get a normal vectors and the height of the first grid on "Nodal BC"

### DISCUSSION THREAD

**Discussion by suqingji** [idaholab/moose#21428] [Q&A Modules: Navier-Stokes] [answered]
Dear,
I want to add a Dirichlet BC to an advection-diffusion problem, but the node values need to be calculated by the normal vector and the height of the first grid.
In fact, IntegratedBC can help me calculate the normal vectors through `_current_side_elem[_qp]->normal()`. Unfortunately, I need to set up a DirichletBC. The NodalBC have no function to compute normal vectors. Even though I add `_normals(_assembly.normals())` in my class. It still could not  compute normal vectors correctly.

tips: 
1. The height of the first grid on "Nodal BC"
![image](https://user-images.githubusercontent.com/61961455/175951185-0f62a606-44df-4c72-98a3-ab966fffee75.png)
2. normal vectors
![image](https://user-images.githubusercontent.com/61961455/175952193-656dd3e1-967d-41c8-8eaa-43564317fbf7.png)

Looking forward to your reply.

**Comment by hugary1995** [comment, marked as answer]
I assume this is FE not FV. Either way, the node normals aren't well defined. You'll need to do some sort of averaging over the connected element edges. There is a NodalNormals aux kernel that you could use as a starting point.

**Reply by GiudGiud** [reply]
see this page of documentation for example
https://mooseframework.inl.gov/source/actions/AddNodalNormalsAction.html

**Reply by suqingji** [reply]
Thank you,
Your advice is very helpful to me.
